zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/internal/JsonSerializers.java

```diff
@@ -0,0 +1,256 @@
+        continue;
+      }
+      switch (parser.currentName()) {
+        case "traceId":
+          result.traceId(parser.getText());
+          break;
+        case "parentId":
+          result.parentId(parser.getText());
+          break;
+        case "id":
+          result.id(parser.getText());
+          break;
+        case "kind":
+          result.kind(Span.Kind.valueOf(parser.getText()));
+          break;
+        case "name":
+          result.name(parser.getText());
+          break;
+        case "timestamp":
+          result.timestamp(parser.getLongValue());
+          break;
+        case "duration":
+          result.duration(parser.getLongValue());
+          break;
+        case "localEndpoint":
+          result.localEndpoint(parseEndpoint(parser));
+          break;
+        case "remoteEndpoint":
+          result.remoteEndpoint(parseEndpoint(parser));
+          break;
+        case "annotations":
+          if (value != JsonToken.START_ARRAY) {
+            throw new IOException("Invalid span, expecting annotations array start, got: " +
+              value);
+          }
+          while (parser.nextToken() != JsonToken.END_ARRAY) {
+            Annotation a = parseAnnotation(parser);
+            result.addAnnotation(a.timestamp(), a.value());
+          }
+          break;
+        case "tags":
+          if (value != JsonToken.START_OBJECT) {
+            throw new IOException("Invalid span, expecting tags object, got: " + value);
+          }
+          while (parser.nextValue() != JsonToken.END_OBJECT) {
+            result.putTag(parser.currentName(), parser.getValueAsString());
+          }
+          break;
+        case "debug":
+          result.debug(parser.getBooleanValue());
+          break;
+        case "shared":
+          result.shared(parser.getBooleanValue());
+          break;
+        default:
+          // Skip
+      }
+    }
+
+    return result.build();
+  }
+
+  static Endpoint parseEndpoint(JsonParser parser) throws IOException {
+    if (!parser.isExpectedStartObjectToken()) {
+      throw new IOException("Not a valid JSON object, start token: " +
+        parser.currentToken());
+    }
+
+    String serviceName = null, ipv4 = null, ipv6 = null;
+    int port = 0;
+
+    while (parser.nextToken() != JsonToken.END_OBJECT) {
+      JsonToken value = parser.nextValue();
+      if (value == JsonToken.VALUE_NULL) {
+        continue;
+      }
+
+      switch (parser.currentName()) {
+        case "serviceName":
+          serviceName = parser.getText();
+          break;
+        case "ipv4":
+          ipv4 = parser.getText();
+          break;
+        case "ipv6":
+          ipv6 = parser.getText();
+          break;
+        case "port":
+          port = parser.getIntValue();
+          break;
+        default:
+          // Skip
+      }
+    }
+
+    if (serviceName == null && ipv4 == null && ipv6 == null && port == 0) return null;
+    return Endpoint.newBuilder()
```

Comment:
I presume this part was a straight port of the old one.. I forget the precedence of ipv4 vs ipv6

Reply:
Yup - by the way you can still see the old code in the benchmarks folder. It calls `.ip` with ipv4 before ipv6, meaning ipv6 has precedence.